### PR TITLE
fix(schedule): bucket publish/unpublish weeks in the restaurant's timezone

### PR DESCRIPTION
**Pull request by jdelgado2002**
Follow-up to #671, which recorded this as tracked non-goal #2.

## Problem

`publish_schedule` and `unpublish_schedule` bucketed shifts with a bare cast:

```sql
AND start_time::date >= p_week_start
AND start_time::date <= p_week_end
```

Casting a `timestamptz` to `date` resolves against the **database session** timezone, not the restaurant's IANA timezone. No migration sets a non-default `TimeZone`, so this ran as UTC.

For an `America/Chicago` restaurant, a Sun 22:00 closing shift has `start_time` = Mon 03:00 UTC — already the next calendar day in UTC, several hours before local midnight. Such a shift landed on the wrong side of the week boundary in both directions:

| Shift (restaurant-local) | UTC day | Old behaviour | Correct |
|---|---|---|---|
| Sun 22:00 on `week_end` | next Mon | wrongly **excluded** | in week |
| Sun 22:00 on `week_start − 1` | Mon = `week_start` | wrongly **included** | prior week |

`Asia/Tokyo` mirrors it in the other direction (Mon 06:00 JST = 21:00 UTC the previous Sunday).

## Fix

Both functions now resolve `COALESCE(r.timezone, 'UTC')` into `v_tz` and bucket with `(s.start_time AT TIME ZONE v_tz)::date` — the pattern already established next door in `get_open_shifts`. An invalid stored IANA name falls back to UTC via an `invalid_parameter_value` handler rather than raising.

`CREATE OR REPLACE FUNCTION` is a full-body rewrite, so `SECURITY DEFINER` is restated, and `SET search_path = public, pg_temp` is added to satisfy the Supabase advisor lint.

### `EXECUTE` privilege boundary

Supabase's default privileges grant `EXECUTE` on public functions to `anon` as well as `authenticated`, so `REVOKE ... FROM PUBLIC` alone is insufficient. Both functions are now explicitly revoked from `PUBLIC, anon` and granted to `authenticated, service_role`. Verified no edge-function caller depends on the anonymous path.

## Added during review

CodeRabbit flagged, at **Critical**, that both functions are `SECURITY DEFINER` and carried **no authorization check**. The `EXECUTE` revoke above stops `anon`, but `authenticated` is a single shared Postgres role — the grant cannot distinguish tenants. Any logged-in user could pass another restaurant's UUID and publish, lock, or unpublish that tenant's week.

**Pre-existing, not introduced here:** `20251123000000` defines both with `SECURITY DEFINER` and no check (lines 222, 265), and nothing redefines them in between. This PR had in fact already *narrowed* it by revoking the default `PUBLIC`/`anon` EXECUTE.

Fixed rather than deferred, because this migration is already restating access control on these two functions — shipping a hardening pass that closes the anonymous half and leaves the cross-tenant half open would be worse than not touching grants at all.

```sql
IF NOT public.user_has_restaurant_access(p_restaurant_id, false) THEN
  RAISE EXCEPTION '...' USING ERRCODE = 'insufficient_privilege';
END IF;
```

`false`, not `true`, is deliberate: the helper's second argument restricts to `owner`/`manager`, but **which roles may publish is a product decision** and the UI's gating is the current authority. Passing `true` would silently change who can publish as a side effect of a timezone fix. Recorded as a tracked non-goal. Only callers are `useSchedulePublish.tsx` via `supabase.rpc()` with a user JWT — no edge function, cron, or `service_role` path loses access.

Also: `publish_schedule`'s `shift_count` now comes from `GET DIAGNOSTICS ... ROW_COUNT` instead of a preceding `COUNT(*)`. Concurrent publishes of the same week would both see the same drafts; the loser updates zero rows but still recorded the stale count. `unpublish_schedule` in the same file already did it this way.

## Tests

New pgTAP suite `supabase/tests/publish_schedule_tz_bucketing.test.sql` — **19 assertions** across three restaurants (`America/Chicago`, `Asia/Tokyo`, and one with an invalid IANA name), exercising a 10pm–2am shift at **both** week boundaries:

```
ok 1  - control shift (Mon 10:00 America/Chicago) is published
ok 2  - Sun 22:00 America/Chicago on week_end is published (03:00 UTC next Monday)
ok 3  - Sun 22:00 America/Chicago before week_start is NOT published (belongs to prior week)
ok 4  - schedule_publications.shift_count counts exactly the two in-week shifts
ok 5  - unpublish clears the Sun 22:00 week_end shift it published
ok 6  - unpublish leaves the prior-week Sun 22:00 shift alone
ok 7  - unpublish_schedule returns 2 (control + week_end closing shift)
ok 8  - Mon 06:00 Asia/Tokyo on week_start is published (21:00 UTC previous Sunday)
ok 9  - Mon 06:00 Asia/Tokyo after week_end is NOT published (21:00 UTC on week_end)
ok 10 - publish_schedule succeeds for a restaurant with an invalid IANA timezone
ok 11 - invalid timezone buckets as UTC rather than raising
ok 12 - anon cannot execute publish_schedule
ok 13 - anon cannot execute unpublish_schedule
ok 14 - authenticated can execute publish_schedule
ok 15 - authenticated can execute unpublish_schedule
ok 16 - a non-member authenticated caller cannot publish another restaurant's week
ok 17 - a non-member authenticated caller cannot unpublish another restaurant's week
ok 18 - the denied unpublish left the target restaurant's published shift untouched
ok 19 - the denied publish recorded no schedule_publications row
```

Assertions 2, 3, 5, 6, 8, 9, 12 and 13 were confirmed RED against the old bodies. The suite is deliberately two-sided: a fix that merely widened the upper bound would still fail assertion 3.

## Verification

- `npm run typecheck` — clean
- `npm run test` — 7957 passed, 2 skipped
- `npm run db:reset && npm run test:db` — **2120 passed, 0 failed**
- `npm run test:tz` — 13 passed
- `npm run lint` — no new findings (the two in `broadcast-open-shifts.spec.ts` predate this branch)

## Not changed

- **No in-body authorization guard.** Neither function checks the caller belongs to `p_restaurant_id`, so any *authenticated* user can still publish another restaurant's week. The `EXECUTE` boundary above closes the anonymous half; the identity check needs its own decision about which roles may publish and should be reviewable on its own. **Tracked follow-up, and the more serious half of the exposure.**
- **No backfill.** A shift wrongly published under the old bucketing stays published until someone republishes that week — a repeatable manager action.
- **`useWeekPublicationStatus`'s browser-local week anchor.** Post-#671 the hook compares full instants, so a viewer *in the restaurant's own timezone* agrees with the corrected RPC. The two diverge only for a cross-timezone viewer — non-goal #1 from #671, not introduced here, but now observable in one more case. Closing it means threading restaurant tz through the shared week anchor.

## Incidental

`tests/e2e/broadcast-open-shifts.spec.ts` carried a comment claiming the RPC "compares local date". That was false when written (it compared the session date, i.e. UTC); corrected to describe restaurant-local bucketing.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Schedule publishing and unpublishing now bucket shifts by the restaurant’s local timezone.
  * Added safe UTC fallback for missing or invalid timezone settings.
  * Improved week-boundary handling across multiple timezones.

* **Security**
  * Restricted schedule operation access to authenticated and service accounts.

* **Tests**
  * Added coverage for timezone edge cases, fallback behavior, and access permissions.
  * Updated end-to-end test guidance for local-time bucketing.

### CODE REVIEW

**Comment by jdelgado2002** [comment]
@coderabbitai review

File: tests/e2e/broadcast-open-shifts.spec.ts [hunk 1]
```diff
@@ -71,7 +71,9 @@ test.describe('Broadcast Open Shifts', () => {
       const mondayStr = fmt(monday);
 
       // Insert a draft shift for Monday at noon local time (avoids timezone edge cases)
-      // The publish_schedule RPC uses start_time::date which compares local date
+      // publish_schedule buckets shifts by the restaurant's IANA timezone
+      // (restaurants.timezone), so noon is unambiguous — it is the same
+      // calendar day in the restaurant's zone and in UTC.
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply tests/e2e/broadcast-open-shifts.spec.ts]: **✅ Agreed** — Correct — the claim was false at offsets of ±12h or beyond (Pacific/Kiritimati at UTC+14 puts local noon at 22:00 UTC the previous day). The UTC relationship was not what the seed relied on in any case: post-fix publish_schedule buckets by the restaurant-local date, so UTC never enters the comparison. Reworded to state what actually holds — noon sits ~12h from either end of the local day, so the browser-local Monday survives any offset difference under ~12h, which covers the America/Chicago default this spec signs up with — and to point at supabase/tests/publish_schedule_tz_bucketing.test.sql, which covers the boundary directly rather than dodging it. Fixed in `d78e9666`.

File: supabase/migrations/20260729120000_publish_schedule_tz_bucketing.sql [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+AS $$
+DECLARE
+  v_shift_count INTEGER;
+  v_publication_id UUID;
+  v_tz TEXT;
+BEGIN
+  -- Resolve the restaurant's IANA zone ONCE, before any use of v_tz below.
+  SELECT r.timezone INTO v_tz
+  FROM public.restaurants r
+  WHERE r.id = p_restaurant_id;
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply supabase/migrations/20260729120000_publish_schedule_tz_bucketing.sql:76]: **✅ Agreed** — Confirmed and fixed. Both functions are SECURITY DEFINER and had no authorization check, so any authenticated caller could publish/lock/unpublish another tenant's shifts by passing that restaurant_id. Verified pre-existing rather than introduced here: 20251123000000 defines both with SECURITY DEFINER and no check (lines 222, 265), and nothing redefines them in between — this PR had already narrowed it by revoking the default PUBLIC/anon EXECUTE. Closed the cross-tenant half here because the migration is already restating access control on these two functions. Guard uses the canonical, production-verified public.user_has_restaurant_access(p_restaurant_id, false), raising insufficient_privilege before any read or write. One deliberate deviation from your suggestion: membership, not manager-role. The helper's second arg restricts to owner/manager, but which roles may publish is a product decision and the UI's gating is the current authority — passing true would silently change who can publish as a side effect of a timezone fix. Recorded as a tracked non-goal in the design doc. Blast radius checked: the only callers are useSchedulePublish.tsx via supabase.rpc() with a user JWT; no edge function, cron, or service_role path invokes either, so nothing loses access. pgTAP 15 -> 19: cross-tenant denial for both functions plus unchanged-state assertions (the denied unpublish leaves Tokyo's published shift untouched; the denied publish inserts no schedule_publications row). Verified non-vacuous — the 42501 comes from the new RAISE, not from RLS. Fixed in `28578f70`.

File: supabase/migrations/20260729120000_publish_schedule_tz_bucketing.sql [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  UPDATE public.shifts s
+  SET
+    is_published = true,
+    locked = true,
+    published_at = NOW(),
+    published_by = auth.uid()
+  WHERE s.restaurant_id = p_restaurant_id
+    AND (s.start_time AT TIME ZONE v_tz)::date >= p_week_start
+    AND (s.start_time AT TIME ZONE v_tz)::date <= p_week_end
+    AND s.is_published = false;
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply supabase/migrations/20260729120000_publish_schedule_tz_bucketing.sql]: **✅ Agreed** — Correct, and the strongest argument for it is that unpublish_schedule in this same file already did it this way (GET DIAGNOSTICS at what was line 160) — publish_schedule was the inconsistent half. Under READ COMMITTED two concurrent publishes of the same week both see the same draft rows; the loser's UPDATE affects zero rows but the pre-update COUNT(*) would still be written into schedule_publications.shift_count, so the publication claims shifts it did not publish. Applied your suggested diff as-is: dropped the preceding COUNT(*) and derived v_shift_count from GET DIAGNOSTICS after the UPDATE, preserving the restaurant, timezone-bucketed date-range, and is_published = false filters. Existing assertion 4 ('schedule_publications.shift_count counts exactly the two in-week shifts') still passes, which confirms the count is unchanged in the uncontended path. Fixed in `28578f70`.

File: supabase/tests/publish_schedule_tz_bucketing.test.sql [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+INSERT INTO restaurants (id, name, timezone) VALUES
+  ('a0000000-0000-0000-0000-00000000c001', 'Chicago Test Restaurant', 'America/Chicago'),
+  ('a0000000-0000-0000-0000-00000000d002', 'Tokyo Test Restaurant',   'Asia/Tokyo'),
+  ('a0000000-0000-0000-0000-00000000e003', 'Bad TZ Test Restaurant',  'Not/AZone')
+ON CONFLICT (id) DO NOTHING;
+
+INSERT INTO employees (id, restaurant_id, name, position) VALUES
+  ('e0000000-0000-0000-0000-00000000c001', 'a0000000-0000-0000-0000-00000000c001', 'Chicago Server', 'Server'),
+  ('e0000000-0000-0000-0000-00000000d002', 'a0000000-0000-0000-0000-00000000d002', 'Tokyo Server',   'Server'),
+  ('e0000000-0000-0000-0000-00000000e003', 'a0000000-0000-0000-0000-00000000e003', 'Bad TZ Server',  'Server');
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply supabase/tests/publish_schedule_tz_bucketing.test.sql]: **✅ Agreed** — Agreed, and it matters more here than the Minor label suggests: the timezone column is the entire subject of this suite, so a restaurant row retained from an earlier run with a different zone would silently invalidate every bucketing assertion rather than failing loudly. Switched both fixtures to ON CONFLICT (id) DO UPDATE assigning the expected values — restaurants sets name and timezone, employees sets restaurant_id, name, and position. The employees insert had no ON CONFLICT clause at all, so on a collision it would have aborted the whole file; that is fixed too. The new user_restaurants membership fixtures use ON CONFLICT (user_id, restaurant_id) DO UPDATE for the same reason. Fixed in `28578f70`.

File: tests/e2e/broadcast-open-shifts.spec.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply tests/e2e/broadcast-open-shifts.spec.ts]: **✅ Agreed** — Unchanged verdict, re-anchored after a rebase onto current main rewrote the SHAs (was d78e9666, now 2b9d5209). The original claim was false at offsets of ±12h or beyond — Pacific/Kiritimati at UTC+14 puts local noon at 22:00 UTC the previous day — so the comment now says only that noon-local is far enough from midnight that the seeded shift keeps its intended local calendar day. Fixed in `2b9d5209`.

File: supabase/migrations/20260729120000_publish_schedule_tz_bucketing.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply supabase/migrations/20260729120000_publish_schedule_tz_bucketing.sql:76]: **✅ Agreed** — Unchanged verdict, re-anchored after the rebase rewrote SHAs (was 28578f70, now cd90d071). Both functions are SECURITY DEFINER and had no caller check, so any authenticated user could pass a foreign restaurant UUID; both now call user_has_restaurant_access(p_restaurant_id, false) and raise 42501. Deliberately membership-only rather than manager-role: which roles may publish is a product decision, recorded as a tracked non-goal in the design doc. pgTAP tests 16-19 cover the denial and prove the state is left untouched. Fixed in `cd90d071`.

File: supabase/migrations/20260729120000_publish_schedule_tz_bucketing.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply supabase/migrations/20260729120000_publish_schedule_tz_bucketing.sql]: **✅ Agreed** — Unchanged verdict, re-anchored after the rebase rewrote SHAs (was 28578f70, now cd90d071). The preceding SELECT COUNT(*) is gone; publish_schedule now uses GET DIAGNOSTICS v_shift_count = ROW_COUNT after the UPDATE, matching what unpublish_schedule already did. That removes the READ COMMITTED race where the losing concurrent publish updates zero rows but records the pre-update count. Fixed in `cd90d071`.

File: supabase/tests/publish_schedule_tz_bucketing.test.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply supabase/tests/publish_schedule_tz_bucketing.test.sql]: **✅ Agreed** — Unchanged verdict, re-anchored after the rebase rewrote SHAs (was 28578f70, now cd90d071). Every fixed-ID fixture now uses ON CONFLICT ... DO UPDATE instead of DO NOTHING — restaurants (name, timezone) and employees (restaurant_id, name, position, which previously had no ON CONFLICT clause at all). This matters more than Minor suggests: a retained row with a stale timezone would silently invalidate the bucketing assertions the suite exists to make. Fixed in `cd90d071`.
